### ISSUE TITLE
Suggested Model (w/ schema and order stored with Series)

### ISSUE THREAD

**Issue by cloudiirain** [Shadowys/hako#26]
This issue is the result of... lots of discussion on Discord. I will try to explain and summarize the best I can.
### What's wrong with the current model?

The current model is essentially: `Project > Series > Volume > Chapter`

Each collection requires its parent as a prerequisite, which is a very rigid model. In real life, a lot of things don't follow this exact hierarchy.

For instance, Web Novels are published online and do not have Volumes. The more natural organization for Web Novels is: `Series > Chapter` (skipping the Volume). We can work around this by assigning all Web Novel chapters to a **DefaultVolume**, but things can get unwieldy considering this isn't the only exception that tends to occur.

For instance, some translators really like to organize their Chapters into Parts. Each chapter contains multiple parts. 

In addition, some Series might be part of a Franchise.

Or some Volumes in a Series might be organized into Arcs.

The brute-force approach to handling all of these extra models is assign DefaultFranchise or DefaultChapter or DefaultArc whenever there is no natural parent. However, this gets ugly quickly (and the API can get very cluttered and complicated). I mean, just imagine having a very thorough model like this:

`Franchise > Series > Arc > Volume > Chapter > Part`

And then try fitting the following case studies to this model:
- A web novel has no franchise, no arcs, no volumes, and no parts.
- A web novel that has arcs but no volumes. (We need to skip the volume model?)

Basically, parsing this can get messy quickly.
### Stripping Things Down to the Essentials

After a long conversation (including Nyar) on discord, the idea we came up with was to figure out what collections were **essential (required)** and which collections were **supplementary (optional)** that we could try implementing on top after securing the essentials..

The simplest we can get is the model every manga aggregator uses:

`Series > Release`

These two collections are essential. Note that manga is organized into volumes as well, but most manga aggregators choose to flatten this hierarchal structure because it's way easier to manage.

Going with this bare-bones hierarchy will work with pretty much everything. Nanodesu's database will work fine with ours, Web Novels won't have any issues, and Light Novels will also work fine (but the structure will be flattened).
### Building on Top of the Essentials

With a little bit of ingenuity, we can define some **supplemental collections** (optional) that build on top of the essentials. 

Let's try implementing an supplemental Volume collection:

```
# Series
- title (of the series)
- extra metadata
- schema (JSON object)

# Release aka Chapter
- title (of the chapter)
- text (of the chapter)
- user_id (of the translator/owner)
- series_id

# Volume
- title (of the volume)
- image (of the volume)
- extra metadata
- series_id
```

Notice how there are very few relationships defined.  The only ones here are (independently): `Series > Release` and `Series > Volume`

**So how would the front-end handle a GET request to display the index page for Series_id=4?**
1. Run three AJAX calls to the API backend?
   - GET `/api/series/<id=4>` -- returns a single series
   - GET `/api/chapter/?where_series_id=4` -- returns a list of releases aka. chapters
   - GET `/api/volume/?where_series_id=4` -- returns a list of volumes
2. Parse Series.schema returned from the first AJAX request.
3. Using the list of Releases and Volumes (acquired from the 2nd and 3rd AJAX calls), the front-end can parse the schema and insert the relevant data at the right position using the object_ids. 

```
{
    "Series.schema": [
        {
            "model": "Volume",
            "id": 213,
            "children": [
                {
                    "model": "Chapter",
                    "id": 321
                },
                {
                    "model": "Chapter",
                    "id": 322
                },
                {
                    "model": "Chapter",
                    "id": 323
                }
            ]
        },
        {
            "model": "Volume",
            "id": 214,
            "children": [
                {
                    "model": "Chapter",
                    "id": 324
                }
            ]
        },
        {
            "model": "Chapter",
            "id": 325
        },
        {
            "model": "Chapter",
            "id": 326
        }
    ]
}
```

Which translates into:
1. Volume 213
   1. Chapter 321
   2. Chapter 322
   3. Chapter 323
2. Volume 214
   1. Chapter 324
3. Chapter 325
4. Chapter 326

Notice how chapter 325+326 is at the root instead of nested within a volume. Using a schema has the advantage in that we don't need to define the relationships between volumes and chapters. Furthermore, we can place Chapters at the root of the schema (to support Web Novels) without going through the process of using default volumes, etc.

**Important**: We will need to strictly validate schemas submitted to the backend. In other words, the front-end can't throw in extra elements that aren't defined or nest volumes within volumes. **We determine the rules** of what constitutes a valid schema.
### What are the advantages of using Series.schema
1. Without a schema, volume and chapter order needs to be defined. We end up needing to update a lot of chapter objects whenever a chapter is re-ordered or inserted.
   - If Series holds the schema, we don't need to store chapter order.
   - Re-ordering chapters only updates the Series object, which makes sense from a object_history standpoint.
2. The models are simple and flexible -- everything hinges on what rules we define for the schema.
   - IMPORTANT: We define the rules! We make the rules! We can make the rules logical! It doesn't have to be as arbitrary as Baka-Tsuki's pages.
   - Simple endpoints for each collection. The relationships/order is stored in one place.

**Comment by cloudiirain** [comment]
Alternative schema:

```
# Series
- title (of the series)
- extra metadata
- children (JSON object schema - ordered list - can have mixed object references)

# Release aka Chapter
- title (of the chapter)
- text (of the chapter)
- user_id (of the translator/owner)
- series_id

# Volume
- title (of the volume)
- image (of the volume)
- extra metadata
- series_id
- children (JSON object schema - ordered list - should only have references to chapters)
```

To make our app interoperable with Nanodesu's backend (and any other apps we work together with), we can do the following:

1. Series = Series
2. Release = Chapter = Page

Add the following field to both models:
```
Series.app_data = {
    "nanodesu_app": {
         // Any custom data that the nanodesu app needs - Black box
     },
    "nandaka_app": {
        // Any custom data that Nandaka's app needs - Black box
    }
}

Release.app_data = {
    "nanodesu_app": {
         // Any custom data that the nanodesu app needs - Black box
     },
    "nandaka_app": {
        // Any custom data that Nandaka's app needs - Black box
    }
}
```

To do this, we'll probably need a system (perhaps a token system) where approved front-end apps can write extra metadata to certain models. This metadata shouldn't be used by anyone except the front-end app... so it's a little bit like a cookie.

The convention would be that approved apps can only write to: `Model.app_data.<app_name> = {}` ...Which is a blackbox dictionary for any addition fields or stuff they need to throw in there. For security reasons, we can probably put a size/length cap on the amount of blackbox data that can be put into app_data.

**Comment by cloudiirain** [comment]
### Footnote 1

Oh, @Shadowys also mentioned something about **Nested Connections**. I don't quite understand what this means, but I think the relationships here can be unidirectional.

In other words, the Series.schema would only contain references to primary key ids. 

Therefore, when a child is updated (e.g. a volume or a chapter), there is no need to update the parent schema, because the id will never change.

In this sense, the relationship is unidirectional. Chapters knows what Series it belongs to, but Series does not know what Chapters belong to it.

If we really wanted a bidirectional relationship, we could do this though:

```
# Series
- title (of the series)
- extra metadata
- schema (JSON object schema - ordered list - mixed object _ids to chapters and volumes)
- volumes - backwards relationship - is a list of volumes that are children
- chapters - backwards relationship - is a list of chapters that are children
```

### Footnote 2

I can empathize with the comment that schemas might be hard to parse. For people who are lazy, maybe we can implement a shortcut API endpoint that flattens the schema:

`/api/series/<id>/?schema_format=chaptersonly`

```
{
	"schema": [
		{
			"model": "Chapter",
			"id": 321
		},
		{
			"model": "Chapter",
			"id": 322
		},
		{
			"model": "Chapter",
			"id": 323
		},
		{
			"model": "Chapter",
			"id": 324
		},
		{
			"model": "Chapter",
			"id": 325
		},
		{
			"model": "Chapter",
			"id": 326
		}
	]
}
```

`/api/series/<id>/?schema_format=volumesonly`

```
{
	"schema": [
		{
			"model": "Volume",
			"id": 213
		},
		{
			"model": "Volume",
			"id": 214
		}
	]
}
```

I mean, honestly, it's really flexible and we can choose to implement schema however we want.

**Comment by cloudiirain** [comment]
```
// Request Series with id==4
{
	"id": 4,
	"title": "Zero no Tsukaima",
	"author": "Yamaguchi Noboru",
	"synopsis": "At the Tristain Magic Academy, her classmates call bumbling witch-in-training Louise Francoise le Blanc de la Valliere “Louise the Zero.” During an important coming-of-age ritual, when each student must summon their own witch’s familiar, Louise confirms her classmates’ opinion by accidentally conjuring up a normal teenage boy from Japan. Whether the two of them like it or not, the laws of magic bind them as master and servant forever! A hilarious novel with romance interwined.",
	"schema": [
		{
			"model": "Volume",
			"id": 32
		},
		{
			"model": "Volume",
			"id": 42
		},
		{
			"model": "Volume",
			"id": 54
		},
		{
			"model": "Chapter", 
			"id": 43153 // This is a bonus chapter that doesn't belong in a volume
		},
		{
			"model": "Chapter", 
			"id": 43153 // This is another bonus chapter that doesn't belong in a volume
		}
	],
	"chapters": [children] // unordered list
	"volumes": [children] // unordered list
	"app_data": {
		"nanodesu_app": {
			// Nanodesu's custom data
		}
	}
}

// Chapters (children of Series)
[
	{
		"id": 43231,
		"title": "Chapter 1 - A New Adventure",
		"text": "lorem ipsum..........",
		"user_id": [foreign key],
		"series_id": [foreign key],
		"volume_id": [can be NULL],
		"app_data": {
			"nanodesu_app": {
				// Nanodesu's custom data
			}
		}
	},
	{
		"id": 43423,
		"title": "Chapter 73 - The Climax",
		"text": "lorem ipsum..........",
		"user_id": [foreign key],
		"series_id": [foreign key],
		"volume_id": [can be NULL],
		"app_data": {
			"nanodesu_app": {
				// Nanodesu's custom data
			}
		}
	},
	[...]
	{
		"id": 43329,
		"title": "Chapter 2 - List of objects is not ordered when returned by mongo",
		"text": "lorem ipsum..........",
		"user_id": [foreign key],
		"series_id": [foreign key],
		"volume_id": [can be NULL],
		"app_data": {
			"nanodesu_app": {
				// Nanodesu's custom data
			}
		}
	},
}

// Volumes (children of Series)
[
	{
		"id": 32,
		"title": "Volume 1",
		"series_id": [foreign key],
		"image": [image]
		"schema": [
			{
				"model": "Chapter",
				"id": 43156
			},
			{
				"model": "Chapter",
				"id": 43156
			},
			{
				"model": "Chapter",
				"id": 43156
			},
			{
				"model": "Chapter",
				"id": 43156
			}
		],
		"chapters": [children] // unordered list
		"app_data": {
			"nanodesu_app": {
				// Nanodesu's custom data
			}
		}
	},
	{
		"id": 42,
		"title": "Volume 2",
		"series_id": [foreign key],
		"image": [image]
		"schema": [
			{
				"model": "Chapter",
				"id": 43195
			}
		],
		"chapters": [children] // unordered list
		"app_data": {
			"nanodesu_app": {
				// Nanodesu's custom data
			}
		}
	}
]
```

**Comment by cloudiirain** [comment]
Continuing the chain of comments (lol). This is a more comprehensive proposed schema. It's just a suggestion so feel free to revise at will:

```
// -------------------------------------------------------------
// Essential Core Collections

// Series Collection (fields derived from major databases)
- id
- title: UNIQUE
- author(s): 
- raw_language: // ex: Jp
- translated_language: // ex: En
- locked: // boolean - meaning whether this series is licensed/DMCA'd
- type: // ex: Light Novel, Web Novel
- artist(s): NULL
- publisher(s): NULL
- desc: NULL
- image: NULL // thumbnail image for the series
- year: NULL // year first chapter/volume was published
- status_in_country_of_origin: NULL // ex: Complete at 8 volumes
- aliases: (implementation still under discussion, otherwise just a list) NULL
- franchise_id: [foreign key] or NULL
- schema: ordered list // describes order of objects at root of the series; needs extensive validation
- app_data: NULL dictionary // for front-end apps like Nanodesu
- releases: [children]
- volumes: [children]

// Release Collection = Page = Part (or Chapter)
- id
- series_id: [foreign key]
- title: UNIQUE in compound index with series_id and volume_id
- text: // translation body
- owner: [foreign key to user]
- collaborator(s): [list of foreign keys to users (or groups) -- needs discussion]
- volume_id: [foreign key] or NULL
- chapter_id: [foreign key] or NULL // if release == part
- arc_id: [foreign key] or NULL

// ModelHistory Collection
- id
- model
- object_id
- timestamp
- user_id
- data: JSON
```

```
// -------------------------------------------------------------
// Supplementary Collections

// Franchise Collection (everything that shares a franchise ID is a related series)
- id
- series: [children]

// Volume Collection
- id
- series_id: [foreign key]
- title: UNIQUE in compound index with series_id
- image: NULL
- desc: NULL
- date_published: NULL
- locked: boolean // putting this at two levels in case admins want to only remove some volumes
- schema: ordered list // describes order of chapters or releases that are children of this volume
- arc_id: NULL
- chapters: [children]
- releases: [children]

// Arc Collection -- Under debate still -- a grouping of volumes or chapters or releases
- id
- series_id
- title: UNIQUE in compound index with series_id
- schema: ordered list // describes order of volumes or chapters or releases that are children of this volume
- volumes: [children]
- chapters: [children]
- releases: [children]

// Chapter Collection -- under debate still - use this model if Release == Part (not a full chapter)
- id
- series_id
- title: UNIQUE in compound index with series_id and volume_id
- schema: ordered list // describes order of releases that are children of this chapter
- volume_id: NULL
- arc_id: NULL
- releases: [children]
```

### Schema Rules:

- Series.schema (root):
    - Valid children: Arc, Volume, Chapter, Release
- Arc.schema:
    - Valid children: Volume, Chapter, Release
- Volume.schema:
    - Valid children: Chapter, Release
- Chapter.schema:
    - Valid children: Release

The point of this schema is that any Model can appear at the root level above them. For instance, Chapters can appear as children of an Arc (like in a Web Novel). 

Nothing can nest itself though (Like, Volumes cannot nest Volumes, and Arcs cannot nest Arcs). Nothing can nest backwards either (like, a Volume cannot have an Arc as children).

The nice thing about the supplementary collections is that they're purely organizational. If the user organizing this project doesn't feel like putting things in Arcs or Parts, they don't have to. Their hierarchy just goes: `Series > Volume > Release`.

You're right that this is a little bit more for the front-end to think about, but with strict rules on the schema, everything is predictable. We can enforce that there are no exceptions on this system. In addition, for lazy front-end programmers, we can just flatten the entire structure and just present it as: `Series > Release`

**Comment by Shadowys** [comment]
I would definitely suggest a unidirectional relation, mainly because maintaining two sync key tables is going to be hard to implement. A singular table will work fine and is adequate. 

I'm also suggesting that the schema can be flattened instead of nested, which also would make parsing it easier, since the parsing algorithm can search linearly instead of transversing through a tree structure. 

However, in this case it may be more appropriate to say that to borrow Nyar's model, each release can have the metadata of "model", which may only equal Arc, Volume, Chapter, Part. The model will have a defined relationship within the schema, and also the guaratee that they cannot nest themselves, nor to their children.

**Comment by cloudiirain** [comment]
I can agree with the unidirectional relation and flattened (linear) schema.

Can you maybe describe the metadata strategy a little bit more? Is it something like this?

```
// Release collection
- id
- series_id: [foreign key]
- title: UNIQUE in compound index with series_id and volume_id
- text: // translation body
- owner: [foreign key to user]
- collaborator(s): [list of foreign keys to users (or groups) -- needs discussion]
- // Isn't all of this below metadata? But separated into different fields?
- arc_id: [foreign key] or NULL
- volume_id: [foreign key] or NULL
- chapter_id: [foreign key] or NUL
```

Because volumes clearly have additional structured data associated with them. In Nyar's model, I think all of it is unstructured in the text body of Page. The metadata of "model" just flags how to process the Page text body for the purposes of their generator.

However, I'm not sure if you're suggesting that we store volume metadata at the level of Release (or Page)?

With an `arc_id` or `volume_id` in the metadata, we can link every release to the collection object that is the parent of this Release. Hence, the Volume Collection can still hold the volume image and ISBN information, etc (and we don't need to store it with releases).

**I think I need to clarify what you mean by a flattened schema.** I can imagine two ways of having a flattened schema:

```
// Option 1: Recursive linear schemas:
SeriesA.schema = [
    <Arc A>,
    <Arc B>,
    <Volume A>,
    <Volume B>
]
ArcA.schema = [
   <Volume C>,
   <Volume D>,
   <Chapter D>
]
VolumeA.schema = [
    <Chapter A>,
    <Chapter B>,
    <Chapter C>
]

// Option 2: Legitimately Flat Schema
SeriesA.schema = [
    <ArcA>,
    <Volume C>, // child of ArcA
    <Chapter 1>, // child of VolumeC, which is a child of ArcA
    <Chapter 2>, // child of VolumeC, which is a child of ArcA
    <Volume D>, // child of ArcA
    <Chapter 3>, // child of VolumeD, which is a child of ArcA
    <Chapter D>, // child of ArcA
    <Arc B>,
    <Volume F>, // child of ArcB
    <Chapter 5>, // child of VolumeF, which is a child of ArcB
    <Chapter 6>, // child of VolumeF, which is a child of ArcB
    <Volume G>, // child of ArcB
    <Chapter 7>, // child of VolumeG, which is a child of ArcB
    <Volume A>, // root, no parent
    <Chapter 8>, // child of VolumeA
    <Chapter 9>, // child of VolumeA
    <Volume B> // root, no parent
    <Chapter 10>, // child of VolumeB
]
```

**Comment by Shadowys** [comment]
For the former, I've forgotten that volumes have additional metadata to them. In that case, we couldn't model them together.

For the latter I'm leaning towards the latter....yeah that doesn't make sense. I think having the series record everything is much preferred?

**Comment by cloudiirain** [comment]
I'm not sure if it's possible to have the series record everything if we do a linear/flat schema (Option 2).

For instance, if I am traversing the Option 2 list, I need to know how of many of the following objects belong to Volume D. For instance, if you pay close attention here:

```
    <ArcA>,
    <Volume C>, // child of ArcA
    <Chapter 1>, // child of VolumeC, which is a child of ArcA
    <Chapter 2>, // child of VolumeC, which is a child of ArcA
    <Volume D>, // child of ArcA
    <Chapter 3>, // child of VolumeD, which is a child of ArcA
    <Chapter D>, // child of ArcA
``
Chapter 3 is a child of Volume D, but Chapter D is not. Someone when parsing this list, we need to be able to differentiate this, which means some additional ordering/metadata needs to be stored with the volumes.

**Comment by Shadowys** [comment]
I was wondering if there a case where chapter may actually only belong to an arc instead of an volume? If there is, then a tree makes parsing through it easier. I was assuming that if volumes were organised under arcs, arcs cannot take chapters as children

**Comment by cloudiirain** [comment]
Well, in reality, it's unlikely that any series would have weird mixed hierarchies like I've been using in  these examples. 

I think most series will have consistent hierarchies:

- `Series > Arc > Chapters > Releases`
- `Franchise > Series > Volumes > Releases`
- `Series > Arc > Releases`

And all variations of this. It's rare that a series might mix up a hierarchy, but I'm not denying the possibility that it might happen (for instance, on some Baka-Tsuki projects, there's random extra tidbits that are translated like an interview with an author, or a drama CD, or something that doesn't belong in a volume).

But yeah, chapters could only belong to an arc if it were a Web Novel that organizes things by arc.

**Comment by Shadowys** [comment]
I see. Then I guess it's much easier for us to maintain the structure through trees instead.

**Comment by cloudiirain** [comment]
I mean, we can make Option 2 work if we make a rule that series can't have mixed object types at any level in the schema. This way, each series would have to be internally consistent with itself.

For instance, if this Series uses Arcs, everything in this Series needs to use the hierarchy `Series > Arc > Volume` and volumes that appear at root (`Series > Volume`) are rejected.

```
// This kind of schema is not allowed
SeriesA.schema = [ // Mixed object types in the schema
    <Arc A>,
    <Arc B>,
    <Volume A>, 
    <Volume B>
]
ArcA.schema = [ // Mixed object types in this schema
   <Volume C>,
   <Volume D>,
   <Chapter D>
]
VolumeA.schema = [ // Single object type -- this is valid
    <Chapter A>,
    <Chapter B>,
    <Chapter C>
]
```

```
// This is valid
SeriesA.schema = [ // Single object type -- this is valid
    <Arc A>,
    <Arc B>,
]
ArcA.schema = [ // Single object type -- this is valid
   <Volume C>,
   <Volume D>,
]
VolumeA.schema = [ // Single object type -- this is valid
    <Chapter A>,
    <Chapter B>,
    <Chapter C>
]
```


```
// This is valid
SeriesA.schema = [ // Single object type -- this is valid
    <Volume A>,
    <Volume B>,
]
VolumeA.schema = [ // Single object type -- this is valid
    <Release A>,
    <Release B>,
    <Release C>
]
```

```
// This is valid
SeriesA.schema = [ // Single object type -- this is valid
    <Chapter A>,
    <Chapter B>,
    <Chapter C>
]
```

This kind of schema might be easier/simpler, if we want to go with this.

If we go with this method, rendering a flat/linear schema would always work.

**Comment by Shadowys** [comment]
I was thinking in the case where you pointed out the interviews and stuff. Those should be classified under chapters/releases instead of volumes, so they are a standalone item.

**Comment by Shadowys** [comment]
A new schema is now available from this commit : 

https://github.com/Shadowys/hako/commit/32386c0104e73cfa08627cb32f4c9ef3e031ad21
